File: solutions/Virtual-Assistant/src/csharp/microsoft.bot.solutions/Skills/SkillDialog.cs
﻿using System;
using System.Collections.Generic;
using System.Globalization;
using System.Linq;
using System.Reflection;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.Bot.Builder;
using Microsoft.Bot.Builder.Azure;
using Microsoft.Bot.Builder.Dialogs;
using Microsoft.Bot.Configuration;
using Microsoft.Bot.Schema;
using Microsoft.Bot.Solutions.Authentication;
using Microsoft.Bot.Solutions.Middleware;
using Microsoft.Bot.Solutions.Proactive;
using Microsoft.Bot.Solutions.Responses;
using Microsoft.Bot.Solutions.TaskExtensions;
using Microsoft.Bot.Solutions.Telemetry;

namespace Microsoft.Bot.Solutions.Skills
{
    public class SkillDialog : ComponentDialog
    {
        // Fields
        private SkillDefinition _skillDefinition;
        private SkillConfigurationBase _skillConfiguration;
        private ResponseManager _responseManager;
        private EndpointService _endpointService;
        private ProactiveState _proactiveState;
        private IBotTelemetryClient _telemetryClient;
        private IBackgroundTaskQueue _backgroundTaskQueue;
        private InProcAdapter _inProcAdapter;
        private IBot _activatedSkill;
        private bool _skillInitialized;
        private bool _useCachedTokens;

        public SkillDialog(SkillDefinition skillDefinition, SkillConfigurationBase skillConfiguration, ProactiveState proactiveState, EndpointService endpointService, IBotTelemetryClient telemetryClient, IBackgroundTaskQueue backgroundTaskQueue, bool useCachedTokens = true)
            : base(skillDefinition.Id)
        {
            _skillDefinition = skillDefinition;
            _skillConfiguration = skillConfiguration;
            _proactiveState = proactiveState;
            _endpointService = endpointService;
            _telemetryClient = telemetryClient;
            _backgroundTaskQueue = backgroundTaskQueue;
            _useCachedTokens = useCachedTokens;

            var supportedLanguages = skillConfiguration.LocaleConfigurations.Keys.ToArray();
            _responseManager = new ResponseManager(supportedLanguages, new SkillResponses());

            AddDialog(new MultiProviderAuthDialog(skillConfiguration));
        }

        protected override async Task<DialogTurnResult> OnBeginDialogAsync(DialogContext innerDc, object options, CancellationToken cancellationToken = default(CancellationToken))
        {
            var skillOptions = (SkillDialogOptions)options;

            // Send parameters to skill in skillBegin event
            var userData = new Dictionary<string, object>();

            if (_skillDefinition.Parameters != null)
            {
                foreach (var parameter in _skillDefinition.Parameters)
                {
                    if (skillOptions.Parameters.TryGetValue(parameter, out var paramValue))
                    {
                        userData.TryAdd(parameter, paramValue);
                    }
                }
            }

            var activity = innerDc.Context.Activity;

            var skillBeginEvent = new Activity(
              type: ActivityTypes.Event,
              channelId: activity.ChannelId,
              from: new ChannelAccount(id: activity.From.Id, name: activity.From.Name),
              recipient: new ChannelAccount(id: activity.Recipient.Id, name: activity.Recipient.Name),
              conversation: new ConversationAccount(id: activity.Conversation.Id),
              name: Events.SkillBeginEventName,
              value: userData);

            // Send event to Skill/Bot
            return await ForwardToSkill(innerDc, skillBeginEvent);
        }

        protected override async Task<DialogTurnResult> OnContinueDialogAsync(DialogContext innerDc, CancellationToken cancellationToken = default(CancellationToken))
        {
            var activity = innerDc.Context.Activity;

            if (innerDc.ActiveDialog?.Id == nameof(MultiProviderAuthDialog))
            {
                // Handle magic code auth
                var result = await innerDc.ContinueDialogAsync(cancellationToken);

                // forward the token response to the skill
                if (result.Status == DialogTurnStatus.Complete && result.Result is ProviderTokenResponse)
                {
                    activity.Type = ActivityTypes.Event;
                    activity.Name = Events.TokenResponseEventName;
                    activity.Value = result.Result as ProviderTokenResponse;
                }
                else
                {
                    return result;
                }
            }

            return await ForwardToSkill(innerDc, activity);
        }

        protected override Task<DialogTurnResult> EndComponentAsync(DialogContext outerDc, object result, CancellationToken cancellationToken)
        {
            return outerDc.EndDialogAsync(result, cancellationToken);
        }

        private async Task InitializeSkill(DialogContext dc)
        {
            try
            {
                IStorage storage;

                if (_skillConfiguration.CosmosDbOptions != null)
                {
                    var cosmosDbOptions = _skillConfiguration.CosmosDbOptions;
                    cosmosDbOptions.CollectionId = _skillDefinition.Name;
                    storage = new CosmosDbStorage(cosmosDbOptions);
                }
                else
                {
                    storage = new MemoryStorage();
                }

                // Initialize skill state
                var userState = new UserState(storage);
                var conversationState = new ConversationState(storage);

                // Create skill instance
                try
                {
                    var skillType = Type.GetType(_skillDefinition.Assembly);

                    // Have to use refined BindingFlags to allow for optional parameters on constructors.
                    _activatedSkill = (IBot)Activator.CreateInstance(
                        skillType,
                        BindingFlags.CreateInstance | BindingFlags.Public | BindingFlags.Instance | BindingFlags.OptionalParamBinding,
                        default(Binder),
                        new object[] { _skillConfiguration, _endpointService, conversationState, userState, _proactiveState, _telemetryClient, _backgroundTaskQueue, true },
                        CultureInfo.CurrentCulture);
                }
                catch (Exception e)
                {
                    var message = $"Skill ({_skillDefinition.Name}) could not be created.";
                    throw new InvalidOperationException(message, e);
                }

                _inProcAdapter = new InProcAdapter
                {
                    // set up skill turn error handling
                    OnTurnError = async (context, exception) =>
                    {
                        await context.SendActivityAsync(_responseManager.GetResponse(SkillResponses.ErrorMessageSkillError));

                        await dc.Context.SendActivityAsync(new Activity(type: ActivityTypes.Trace, text: $"Skill Error: {exception.Message} | {exception.StackTrace}"));

                        // Log exception in AppInsights
                        _telemetryClient.TrackExceptionEx(exception, context.Activity);
                    },
                    BackgroundTaskQueue = _backgroundTaskQueue
                };

                _inProcAdapter.Use(new EventDebuggerMiddleware());

                // change this to use default locale from appsettings when we have dependency injection
                var locale = "en-us";
                if (!string.IsNullOrWhiteSpace(dc.Context.Activity.Locale))
                {
                    locale = dc.Context.Activity.Locale;
                }

                _inProcAdapter.Use(new SetLocaleMiddleware(locale));

                _inProcAdapter.Use(new AutoSaveStateMiddleware(userState, conversationState));
                _skillInitialized = true;
            }
            catch
            {
                // something went wrong initializing the skill, so end dialog cleanly and throw so the error is logged
                _skillInitialized = false;
                await dc.EndDialogAsync();
                throw;
            }
        }

        private async Task<DialogTurnResult> ForwardToSkill(DialogContext innerDc, Activity activity)
        {
            try
            {
                if (!_skillInitialized)
                {
                    await InitializeSkill(innerDc);
                }

                _inProcAdapter.ProcessActivity(activity, async (skillContext, ct) =>
                {
                    await _activatedSkill.OnTurnAsync(skillContext);
                }, async (activities) =>
                {
                    foreach (var response in activities)
                    {
                        await innerDc.Context.Adapter.ContinueConversationAsync(_endpointService.AppId, response.GetConversationReference(), CreateCallback(response), default(CancellationToken));
                    }
                }).Wait();

                var queue = new List<Activity>();
                var endOfConversation = false;
                var skillResponse = _inProcAdapter.GetNextReply();

                while (skillResponse != null)
                {
                    if (skillResponse.Type == ActivityTypes.EndOfConversation)
                    {
                        endOfConversation = true;
                    }
                    else if (skillResponse?.Name == Events.TokenRequestEventName)
                    {
                        // Send trace to emulator
                        await innerDc.Context.SendActivityAsync(new Activity(type: ActivityTypes.Trace, text: $"<--Received a Token Request from a skill"));

                        if (!_useCachedTokens)
                        {
                            var adapter = innerDc.Context.Adapter as BotFrameworkAdapter;
                            var tokens = await adapter.GetTokenStatusAsync(innerDc.Context, innerDc.Context.Activity.From.Id);

                            foreach (var token in tokens)
                            {
                                await adapter.SignOutUserAsync(innerDc.Context, token.ConnectionName, innerDc.Context.Activity.From.Id, default(CancellationToken));
                            }
                        }

                        var authResult = await innerDc.BeginDialogAsync(nameof(MultiProviderAuthDialog));

                        if (authResult.Result?.GetType() == typeof(ProviderTokenResponse))
                        {
                            var tokenEvent = skillResponse.CreateReply();
                            tokenEvent.Type = ActivityTypes.Event;
                            tokenEvent.Name = Events.TokenResponseEventName;
                            tokenEvent.Value = authResult.Result as ProviderTokenResponse;

                            return await ForwardToSkill(innerDc, tokenEvent);
                        }
                        else
                        {
                            return authResult;
                        }
                    }
                    else
                    {
                        if (skillResponse.Type == ActivityTypes.Trace)
                        {
                            // Write out any trace messages from the skill to the emulator
                            await innerDc.Context.SendActivityAsync(skillResponse);
                        }
                        else
                        {
                            queue.Add(skillResponse);
                        }
                    }

                    skillResponse = _inProcAdapter.GetNextReply();
                }

                // send skill queue to User
                if (queue.Count > 0)
                {
                    // if the conversation id from the activity is the same as the context activity, it's reactive message
                    await innerDc.Context.SendActivitiesAsync(queue.ToArray());
                }

                // handle ending the skill conversation
                if (endOfConversation)
                {
                    await innerDc.Context.SendActivityAsync(new Activity(type: ActivityTypes.Trace, text: $"<--Ending the skill conversation"));

                    return await innerDc.EndDialogAsync();
                }
                else
                {
                    return EndOfTurn;
                }
            }
            catch
            {
                // something went wrong forwarding to the skill, so end dialog cleanly and throw so the error is logged.
                // NOTE: errors within the skill itself are handled by the OnTurnError handler on the adapter.
                await innerDc.EndDialogAsync();
                throw;
            }
        }

        private BotCallbackHandler CreateCallback(Activity activity)
        {
            return async (turnContext, token) =>
            {
                EnsureActivity(activity);

                // Send back the activities in the proactive context
                await turnContext.SendActivityAsync(activity, token);
            };
        }

        /// <summary>
        /// Ensure the activity objects are correctly set for proactive messages
        /// There is known issues about not being able to send these messages back
        /// correctly if the properties are not set in a certain way.
        /// </summary>
        /// <param name="activity">activity that's being sent out.</param>
        private void EnsureActivity(Activity activity)
        {
            if (activity != null)
            {
                if (activity.From != null)
                {
                    activity.From.Name = "User";
                    activity.From.Properties["role"] = "user";
                }

                if (activity.Recipient != null)
                {
                    activity.Recipient.Id = "1";
                    activity.Recipient.Name = "Bot";
                    activity.Recipient.Properties["role"] = "bot";
                }
            }
        }

        private class Events
        {
            public const string SkillBeginEventName = "skillBegin";
            public const string TokenRequestEventName = "tokens/request";
            public const string TokenResponseEventName = "tokens/response";
        }
    }
}

File: solutions/Virtual-Assistant/src/typescript/bot-solution/src/skills/skillDialog.ts
/* tslint:disable:no-any */
import { Activity, ActivityTypes, AutoSaveStateMiddleware, BotTelemetryClient, ConversationState,
    MemoryStorage, Storage, TurnContext, UserState } from 'botbuilder';
import { CosmosDbStorage, CosmosDbStorageSettings } from 'botbuilder-azure';
import { ComponentDialog, Dialog, DialogContext, DialogTurnResult, DialogTurnStatus } from 'botbuilder-dialogs';
import { IEndpointService } from 'botframework-config';
import { IProviderTokenResponse, isProviderTokenResponse, MultiProviderAuthDialog } from '../authentication';
import { ActivityExtensions } from '../extensions';
import { EventDebuggerMiddleware, SetLocaleMiddleware, TelemetryExtensions } from '../middleware';
import { CommonResponses } from '../resources';
import { ResponseManager } from '../responses/responseManager';
import { InProcAdapter } from './inProcAdapter';
import { SkillConfigurationBase } from './skillConfigurationBase';
import { SkillDefinition } from './skillDefinition';
import { ISkillDialogOptions } from './skillDialogOptions';

export class SkillDialog extends ComponentDialog {
    // Fields
    private readonly skillDefinition: SkillDefinition;
    private readonly skillConfiguration: SkillConfigurationBase;
    private readonly responseManager: any; // ResponseManager;
    private readonly endpointService: IEndpointService;
    private readonly useCachedTokens: boolean;
    private inProcAdapter: InProcAdapter = new InProcAdapter();
    private activatedSkill: any; // IBot;
    private skillInitialized: boolean;

    constructor(skillDefinition: SkillDefinition,
                skillConfiguration: SkillConfigurationBase,
                endpointService: IEndpointService,
                telemetryClient: BotTelemetryClient,
                useCachedTokens: boolean = true) {
        super(skillDefinition.id);

        this.skillDefinition = skillDefinition;
        this.skillConfiguration = skillConfiguration;
        this.endpointService = endpointService;
        this.telemetryClient = telemetryClient;
        this.useCachedTokens = useCachedTokens;

        this.skillInitialized = false;

        const supportedLanguages: string[] = Array.from(skillConfiguration.localeConfigurations.keys());
        this.responseManager = new ResponseManager([CommonResponses], supportedLanguages);

        this.addDialog(new MultiProviderAuthDialog(skillConfiguration));
    }

    protected onBeginDialog(innerDC: DialogContext, options?: object): Promise<DialogTurnResult> {
        const skillOptions: ISkillDialogOptions = <ISkillDialogOptions> options;

        // Send parameters to skill in skillBegin event
        const userData: Map<string, Object> = new Map();

        if (this.skillDefinition.parameters) {
            this.skillDefinition.parameters.forEach((parameter: string) => {
                if (skillOptions.parameters && skillOptions.parameters.has(parameter)) {
                    userData.set(parameter, skillOptions.parameters.get(parameter) || '');
                }
            });
        }

        const activity: Activity = innerDC.context.activity;

        const skillBeginEvent: Partial<Activity> = {
            type: ActivityTypes.Event,
            channelId: activity.channelId,
            from: activity.from,
            recipient: activity.recipient,
            conversation: activity.conversation,
            name: Events.skillBeginEventName,
            value: userData
        };

        // Send event to Skill/Bot
        return this.forwardToSkill(innerDC, skillBeginEvent);
    }

    protected async onContinueDialog(innerDC: DialogContext): Promise<DialogTurnResult> {
        const activity: Activity = innerDC.context.activity;

        if (innerDC.activeDialog && innerDC.activeDialog.id === MultiProviderAuthDialog.name) {
            // Handle magic code auth
            const result: DialogTurnResult = await innerDC.continueDialog();

            // forward the token response to the skill
            if (result.status === DialogTurnStatus.complete && isProviderTokenResponse(result.result)) {
                activity.type = ActivityTypes.Event;
                activity.name = Events.tokenResponseEventName;
                activity.value = <IProviderTokenResponse> result.result;
            } else {
                return result;
            }
        }

        return this.forwardToSkill(innerDC, activity);
    }

    protected endComponent(outerDC: DialogContext, result: any): Promise<DialogTurnResult> {
        return outerDC.endDialog(result);
    }

    private async initializeSkill(dc: DialogContext): Promise<void> {
        try {
            let storage: Storage;

            if (this.skillConfiguration.cosmosDbOptions) {
                const options: CosmosDbStorageSettings = this.skillConfiguration.cosmosDbOptions;
                options.collectionId = this.skillDefinition.name;
                storage = new CosmosDbStorage(options);
            } else {
                storage = new MemoryStorage();
            }

            // Initialize skill state
            const userState: UserState = new UserState(storage);
            const conversationState: ConversationState = new ConversationState(storage);

            // Create skill instance
            try {
                // PENDING: create skill
                this.activatedSkill = undefined;
                throw new Error('Creation of skills not implemented');

            } catch (error) {
                const message: string = `Skill '${this.skillDefinition.name}' could not be created. (${error})`;
                throw new Error(message);
            }

            // set up skill turn error handling
            this.inProcAdapter = new InProcAdapter();

            this.inProcAdapter.onTurnError = async (context: TurnContext, error: Error): Promise<void> => {
                await context.sendActivity(this.responseManager.getResponse(CommonResponses.errorMessage_SkillError));

                const traceActivity: Partial<Activity> = {
                    type: ActivityTypes.Trace,
                    text: `Skill error: ${error.message} | ${error.stack}`
                };

                await dc.context.sendActivity(traceActivity);

                // Log exception in AppInsights
                TelemetryExtensions.trackExceptionEx(this.telemetryClient, error, context.activity);
            };

            this.inProcAdapter.use(new EventDebuggerMiddleware());
            // PENDING: Apply this fix from C# https://github.com/Microsoft/AI/pull/878
            this.inProcAdapter.use(new SetLocaleMiddleware(dc.context.activity.locale || 'en-us'));
            this.inProcAdapter.use(new AutoSaveStateMiddleware(userState, conversationState));

            this.skillInitialized = true;

        } catch (error) {
            // something went wrong initializing the skill, so end dialog cleanly and throw so the error is logged
            this.skillInitialized = false;
            await dc.endDialog();
            throw error;
        }
    }

    private async forwardToSkill(innerDc: DialogContext, activity: Partial<Activity>): Promise<DialogTurnResult> {
        try {
            if (!this.skillInitialized) {
                await this.initializeSkill(innerDc);
            }

            await this.inProcAdapter.processActivity(activity, async (skillContext: TurnContext): Promise<void> => {
                await this.activatedSkill.onTurn(skillContext);
            });

            const queue: Activity[] = [];
            let endOfConversation: boolean = false;
            let skillResponse: Partial<Activity> | undefined = this.inProcAdapter.getNextReply();

            while (skillResponse) {
                if (skillResponse.type === ActivityTypes.EndOfConversation) {
                    endOfConversation = true;
                } else if (skillResponse.name === Events.tokenRequestEventName) {
                    // Send trace to emulator
                    const traceActivity: Partial<Activity> = {
                        type: ActivityTypes.Trace,
                        text: '<--Received a Token Request from a skill'
                    };

                    await innerDc.context.sendActivity(traceActivity);

                    /* Not implemented
                    if (!this.useCachedTokens) {
                        const adapter: BotFrameworkAdapter = <BotFrameworkAdapter> innerDc.context.adapter;
                        const tokens: TokenStatus[] = adapter.GetTokenStatus(innerDc.context, innerDc.context.activity.from.id);

                        tokens.forEach(async (token: TokenStatus) => {
                            await adapter.signOutUser(innerDc.context, token.connectionName, innerDc.context.activity.from.id);
                        });
                    }
                    */

                    const authResult: DialogTurnResult = await innerDc.beginDialog(MultiProviderAuthDialog.name);
                    if (isProviderTokenResponse(authResult.result)) {
                        const tokenEvent: Activity = ActivityExtensions.createReply(<Activity>skillResponse);
                        tokenEvent.type = ActivityTypes.Event;
                        tokenEvent.name = Events.tokenResponseEventName;
                        tokenEvent.value = <IProviderTokenResponse> authResult.result;

                        return await this.forwardToSkill(innerDc, tokenEvent);

                    } else {
                        return authResult;
                    }
                } else {
                    if (skillResponse.type === ActivityTypes.Trace) {
                        // Write out any trace messages from the skill to the emulator
                        await innerDc.context.sendActivity(skillResponse);
                    } else {
                        queue.push(<Activity>skillResponse);
                    }
                }

                skillResponse = this.inProcAdapter.getNextReply();
            }

            // send skill queue to User
            if (queue.length > 0) {
                const firstActivity: Activity = queue[0];
                if (firstActivity.conversation.id === innerDc.context.activity.conversation.id) {
                    // if the conversation id from the activity is the same as the context activity, it's reactive message
                    await innerDc.context.sendActivities(queue);
                } else {
                    // if the conversation id from the activity is different from the context activity, it's proactive message
                    await innerDc.context.adapter.continueConversation(
                        TurnContext.getConversationReference(firstActivity),
                        this.createCallback(queue));
                }
            }

            // handle ending the skill conversation
            if (endOfConversation) {
                const trace: Partial<Activity> = {
                    type: ActivityTypes.Trace,
                    text: '<--Ending the skill conversation'
                };

                await innerDc.context.sendActivity(trace);

                return await innerDc.endDialog();
            } else {
                return Dialog.EndOfTurn;
            }
        } catch (error) {
            // something went wrong forwarding to the skill, so end dialog cleanly and throw so the error is logged.
            // NOTE: errors within the skill itself are handled by the OnTurnError handler on the adapter.
            await innerDc.endDialog();
            throw error;
        }
    }

    private createCallback(activities: Activity[]): (context: TurnContext) => Promise<void> {
        return async (turnContext: TurnContext): Promise<void> => {
            // Send back the activities in the proactive context
            await turnContext.sendActivities(activities);
        };
    }
}

namespace Events {
    export const skillBeginEventName: string = 'skillBegin';
    export const tokenRequestEventName: string = 'tokens/request';
    export const tokenResponseEventName: string = 'tokens/response';
}
